vizro-core/src/vizro/managers/_data_manager.py

```diff
@@ -1,126 +1,282 @@
 """The data manager handles access to all DataFrames used in a Vizro app."""
 
-from typing import Callable, Dict, Union
+from __future__ import annotations
+
+import warnings
+
+from contextlib import suppress
+
+import logging
+import os
+from typing import Callable, Dict, Optional, Union
 
 import pandas as pd
+import wrapt
+from flask_caching import Cache
 
 from vizro.managers._managers_utils import _state_modifier
 
-# Really ComponentID and DatasetName should be NewType and not just aliases but then for a user's code to type check
+# TODO: test manually and write tests:
+# * new error messages that are raised
+# Only need to test str/non-str static dataframe argument throughout unit tests in general.
+# And test dynamic datasets by themselves
+
+
+#####################
+# Just for the purposes of easily seeing the right debug messages. Remove before merging to main and revert to this:
+# logger = logging.getLogger(__name__)
+class PrefixAdapter(logging.LoggerAdapter):
+    def process(self, msg, kwargs):
+        return f"[DATA MANAGER] {msg}", kwargs
+
+
+logger = PrefixAdapter(logging.getLogger(__name__), {})
+logger.setLevel(logging.DEBUG)
+#####################
+
+# Really ComponentID and DataSourceName should be NewType and not just aliases but then for a user's code to type check
 # correctly they would need to cast all strings to these types.
+# TODO: remove these type aliases once have moved component to data mapping to models
 ComponentID = str
-DatasetName = str
-pd_LazyDataFrame = Callable[[], pd.DataFrame]
+DataSourceName = str
+pd_DataFrameCallable = Callable[[], pd.DataFrame]
```

Comment:
Yes! We will in due course have something very much like what you suggest actually so that you can e.g. specify from the frontend to load only certain columns of a data source. Haven't worked out all the details yet, but basically:
* `_DynamicData.load` would take `*args, *kwargs`
* `memoize` is already setup to handle these and caches based on the value of those arguments. So it can hold multiple caches for one dataset, each with a different set of arguments. (This is basically what memoize means. At the moment we don't have any argument-depend cache because we don't specify any arguments, but that's really what memoize is designed to do)
* possibly `CapturedCallable` is useful here to have a set of arguments that can be overridden at run time

> Also, should the pd_DataFrameCallable type be: `Callable[Any, pd.DataFrame]`?

In future when we add the above functionality, yes. But for now, no - it can't take any arguments.